### PR TITLE
custom function registration using action url from backend

### PR DESCRIPTION
**Pull request by barshat7**
custom function registration using action url from backend

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java [hunk 1]
```diff
@@ -373,4 +373,26 @@ default void visit(Consumer<ComponentExporter> callback) throws Exception {}
...
     }
+
+    /**
+     * Returns the url from where the custom functions should be registered
+     * 
+     * @return custom function registration url
+     * @since com.adobe.cq.forms.core.components.models.form 5.7.6
+     */
+    @JsonIgnore
+    default String getCustomFunctionAction() {
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:384]: Please refer to the implementations of getAction and getDataUrl; there should be no property for this in JCR. For getPrefillAction, let's reuse the existing data URL implementation and rename it to getCustomFunctionURL.

File: ui.frontend/src/utils.js [hunk 2]
```diff
@@ -309,14 +327,15 @@ class Utils {
             } else {
                 const _formJson = await HTTPAPILayer.getFormDefinition(_path, _pageLang);
                 console.debug("fetched model json", _formJson);
-                await this.registerCustomFunctions(_formJson.id);
```
Comment (rismehta) [inline ui.frontend/src/utils.js:312]: We could improve the existing function implementation to incorporate the following logic,
Check if fd:customFunctionsUrl exists, if yes, use it, else fallback to the existing logic

let's not create a new method

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

File: ui.frontend/src/utils.js (hunk 2 above)
Comment (barshat7) [reply ui.frontend/src/utils.js:312]: The intention to create a new function was not to change the signature of the existing one. I see the existing function as exposed and documented.

Comment (rismehta) [reply ui.frontend/src/utils.js:312]: @barshat7 You were right, this function is a public API, https://opensource.adobe.com/aem-core-forms-components/module-FormView-HTTPAPILayer.html

Can we deprecate this API and create a new API with proper JS doc annotation in each API

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

File: ui.frontend/src/utils.js [hunk 3]
```diff
@@ -286,6 +287,30 @@ class Utils {
         }
     }
 
+    /**
+     * Registers custom functions from clientlibs.
+     * @param {object} formJson - The form ID.
```
Comment (rismehta) [inline ui.frontend/src/utils.js:292]: Please update js doc since these are auto published.

**Pull request merged**